File: src/main/java/org/example/kalkulationsprogramm/service/BestellungPdfService.java
package org.example.kalkulationsprogramm.service;

import com.lowagie.text.*;
import com.lowagie.text.Font;
import com.lowagie.text.Image;
import com.lowagie.text.pdf.PdfPCell;
import com.lowagie.text.pdf.PdfPTable;
import com.lowagie.text.pdf.PdfWriter;
import lombok.AllArgsConstructor;
import org.example.kalkulationsprogramm.dto.Bestellung.BestellungResponseDto;
import org.example.kalkulationsprogramm.repository.SchnittbilderRepository;
import org.example.kalkulationsprogramm.service.einkauf.EinkaufPdfPositionsRenderer;
import org.springframework.stereotype.Service;

import java.awt.*;
import java.io.IOException;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

@Service
@AllArgsConstructor
public class BestellungPdfService {

    private final BestellungService bestellungService;
    private final SchnittbilderRepository schnittbilderRepository;
    private final DateiSpeicherService dateiSpeicherService;
    private final FirmeninformationService firmeninformationService;

    private static final byte[] NO_IMAGE = new byte[0];
    private final Map<String, byte[]> schnittbildIconCache = new ConcurrentHashMap<>();

    public Path generatePdfForLieferant(Long lieferantId) {
        List<BestellungResponseDto> alle = bestellungService.findeOffeneBestellungen();
        List<BestellungResponseDto> items = alle.stream()
                .filter(b -> lieferantId == null ? b.getLieferantId() == null : lieferantId.equals(b.getLieferantId()))
                .collect(Collectors.toList());
        Map<String, List<BestellungResponseDto>> byProjekt = items.stream()
                .collect(Collectors.groupingBy(BestellungResponseDto::getProjektName, Collectors.toList()));
        try {
            Path dir = Path.of("uploads");
            Files.createDirectories(dir);
            Path temp = Files.createTempFile(dir, "bestellung-", ".pdf");
            Document doc = new Document(PageSize.A4.rotate());
            PdfWriter writer = PdfWriter.getInstance(doc, Files.newOutputStream(temp));
            writer.setCompressionLevel(0);
            doc.open();

            addCompanyLogo(doc);
            doc.add(new Paragraph(" "));
            String infoText = "Bitte stellen Sie Rechnungen separat pro Auftrag aus. Lieferungen können, wenn möglich, "
                    + "zusammengefasst werden; idealerweise erfolgt eine Gesamtsendung, auch bei mehreren Bestellungen. "
                    + "Die benötigten Meter je Profil entnehmen Sie der Anfrage. Bitte optimieren Sie die Zuschnitte auf Ihre Lagerlängen.";
            Paragraph info = new Paragraph(infoText, FontFactory.getFont(FontFactory.HELVETICA, 10));
            info.setSpacingAfter(15f);
            doc.add(info);

            Font titleFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD, 16, new Color(204, 0, 0));
            Font headerFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD, 10, Color.WHITE);
            Font cellFont = FontFactory.getFont(FontFactory.HELVETICA, 8);
            Color headerBg = new Color(204, 0, 0);
            Color altBg = new Color(245, 245, 245);
            for (Map.Entry<String, List<BestellungResponseDto>> entry : byProjekt.entrySet()) {
                String heading = "Bauvorhaben: " + entry.getKey();
                Paragraph title = new Paragraph(heading, titleFont);
                title.setSpacingAfter(15f);
                doc.add(title);
                PdfPTable table = new PdfPTable(new float[] { 2f, 2f, 2f, 2f, 3f, 1.2f, 1.2f, 1.2f, 2f, 2f, 2f, 1f });
                table.setWidthPercentage(100);
                String[] headers = { "Projektnummer", "Kunde", "Artikelnummer", "Produkt", "Produkttext", "Form",
                        "Winkel L", "Winkel R", "Kommentar", "Werkstoff", "Kategorie", "Menge" };
                for (String h : headers) {
                    PdfPCell cell = new PdfPCell(new Phrase(h, headerFont));
                    cell.setBackgroundColor(headerBg);
                    table.addCell(cell);
                }
                boolean alternate = false;
                for (BestellungResponseDto b : entry.getValue()) {
                    Color bg = alternate ? altBg : Color.WHITE;
                    table.addCell(makeCell(b.getProjektNummer(), cellFont, bg));
                    table.addCell(makeCell(b.getKundenName(), cellFont, bg));
                    table.addCell(makeCell(b.getExterneArtikelnummer(), cellFont, bg));
                    table.addCell(makeCell(b.getProduktname(), cellFont, bg));
                    table.addCell(makeCell(b.getProdukttext(), cellFont, bg));
                    table.addCell(makeCutCell(b.getSchnittForm(), cellFont, bg));
                    table.addCell(makeCell(b.getAnschnittWinkelLinks(), cellFont, bg));
                    table.addCell(makeCell(b.getAnschnittWinkelRechts(), cellFont, bg));
                    table.addCell(makeCell(b.getKommentar(), cellFont, bg));
                    table.addCell(makeCell(b.getWerkstoffName(), cellFont, bg));
                    table.addCell(makeCell(b.getKategorieName(), cellFont, bg));
                    table.addCell(makeCell(formatMenge(b), cellFont, bg));
                    alternate = !alternate;
                }
                doc.add(table);
                doc.add(new Paragraph(" ", cellFont));
            }
            doc.close();
            appendSchnittbildReferenz(temp, dir);
            return temp.toAbsolutePath();
        } catch (IOException e) {
            throw new RuntimeException("PDF generation failed", e);
        }
    }

    public Path generatePdfForProjekt(Long projektId) {
        List<BestellungResponseDto> alle = bestellungService.findeOffeneBestellungen();
        List<BestellungResponseDto> items = alle.stream()
                .filter(b -> projektId.equals(b.getProjektId()))
                .filter(b -> b.getRootKategorieId() != null && b.getRootKategorieId() == 1)
                .collect(Collectors.toList());
        Map<String, List<BestellungResponseDto>> byKat = items.stream()
                .collect(Collectors.groupingBy(b -> b.getKategorieName() != null ? b.getKategorieName() : "",
                        Collectors.toList()));
        try {
            Path dir = Path.of("uploads");
            Files.createDirectories(dir);
            Path temp = Files.createTempFile(dir, "bestellung-", ".pdf");
            Document doc = new Document(PageSize.A4.rotate());
            PdfWriter writer = PdfWriter.getInstance(doc, Files.newOutputStream(temp));
            writer.setCompressionLevel(0);
            doc.open();

            addCompanyLogo(doc);
            doc.add(new Paragraph(" "));
            String infoText = "Bitte stellen Sie Rechnungen separat pro Auftrag aus. "
                    + "Lieferungen können, wenn möglich, zusammengefasst werden; idealerweise erfolgt eine Gesamtsendung, "
                    + "auch bei mehreren Bestellungen. Die benötigten Meter je Profil entnehmen Sie der Anfrage. "
                    + "Bitte optimieren Sie die Zuschnitte auf Ihre Lagerlängen.";
            Paragraph info = new Paragraph(infoText, FontFactory.getFont(FontFactory.HELVETICA, 10));
            info.setSpacingAfter(15f);
            doc.add(info);

            Font titleFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD, 16, new Color(204, 0, 0));
            Font headerFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD, 10, Color.WHITE);
            Font cellFont = FontFactory.getFont(FontFactory.HELVETICA, 8);
            Color headerBg = new Color(204, 0, 0);
            Color altBg = new Color(245, 245, 245);

            for (Map.Entry<String, List<BestellungResponseDto>> entry : byKat.entrySet()) {
                String heading = entry.getKey();
                Paragraph title = new Paragraph(heading, titleFont);
                title.setSpacingAfter(15f);
                doc.add(title);
                PdfPTable table = new PdfPTable(new float[] { 2f, 3f, 3f, 1.2f, 1.2f, 1.2f, 2f, 2f, 1f });
                table.setWidthPercentage(100);
                String[] headers = { "Artikelnummer", "Produkt", "Produkttext", "Form", "Winkel L", "Winkel R",
                        "Werkstoff", "Kommentar", "Menge" };
                for (String h : headers) {
                    PdfPCell cell = new PdfPCell(new Phrase(h, headerFont));
                    cell.setBackgroundColor(headerBg);
                    table.addCell(cell);
                }
                boolean alternate = false;
                for (BestellungResponseDto b : entry.getValue()) {
                    Color bg = alternate ? altBg : Color.WHITE;
                    table.addCell(makeCell(b.getExterneArtikelnummer(), cellFont, bg));
                    table.addCell(makeCell(b.getProduktname(), cellFont, bg));
                    table.addCell(makeCell(b.getProdukttext(), cellFont, bg));
                    table.addCell(makeCutCell(b.getSchnittForm(), cellFont, bg));
                    table.addCell(makeCell(b.getAnschnittWinkelLinks(), cellFont, bg));
                    table.addCell(makeCell(b.getAnschnittWinkelRechts(), cellFont, bg));
                    table.addCell(makeCell(b.getWerkstoffName(), cellFont, bg));
                    table.addCell(makeCell(b.getKommentar(), cellFont, bg));
                    table.addCell(makeCell(formatMenge(b), cellFont, bg));
                    alternate = !alternate;
                }
                doc.add(table);
                doc.add(new Paragraph(" ", cellFont));
            }
            doc.close();
            // Statische Schnittbild-PDF-Seiten am Ende anhängen
            appendSchnittbildReferenz(temp, dir);
            return temp.toAbsolutePath();
        } catch (IOException e) {
            throw new RuntimeException("PDF generation failed", e);
        }
    }

    private void appendSchnittbildReferenz(Path original, Path directory) throws IOException {
        try (var reference = BestellungPdfService.class
                .getResourceAsStream("/static/Schnittbilder_Formstahl_Kuhn Copy.pdf")) {
            if (reference == null) return;
            Path merged = Files.createTempFile(directory, "bestellung-merged-", ".pdf");
            var mainReader = new com.lowagie.text.pdf.PdfReader(Files.readAllBytes(original));
            var referenceReader = new com.lowagie.text.pdf.PdfReader(reference);
            try {
                var mergedDocument = new com.lowagie.text.Document();
                var copy = new com.lowagie.text.pdf.PdfCopy(mergedDocument, Files.newOutputStream(merged));
                mergedDocument.open();
                for (int i = 1; i <= mainReader.getNumberOfPages(); i++) {
                    copy.addPage(copy.getImportedPage(mainReader, i));
                }
                for (int i = 1; i <= referenceReader.getNumberOfPages(); i++) {
                    copy.addPage(copy.getImportedPage(referenceReader, i));
                }
                mergedDocument.close();
            } finally {
                mainReader.close();
                referenceReader.close();
            }
            Files.move(merged, original, java.nio.file.StandardCopyOption.REPLACE_EXISTING);
        }
    }

    /**
     * Fuegt das hochgeladene Firmenlogo oben im PDF ein. Ist kein Logo hinterlegt
     * (oder die Datei fehlt), wird das Logo weggelassen – es gibt bewusst
     * keinen Fallback auf ein Software-Logo, damit niemals ein fremdes Logo
     * auf Handwerker-Dokumenten erscheint.
     */
    private void addCompanyLogo(Document doc) throws DocumentException {
        Image logo = firmeninformationService.loadLogoImage();
        if (logo == null) {
            return;
        }
        logo.scaleToFit(150, 70);
        doc.add(logo);
    }

    private PdfPCell makeCutCell(String form, Font font, Color bg) {
        if (form == null || form.isBlank()) {
            return EinkaufPdfPositionsRenderer.schnittZelle(form, font, bg, null);
        }
        byte[] bytes = loadSchnittbildIcon(form);
        return EinkaufPdfPositionsRenderer.schnittZelle(form, font, bg, bytes);
    }

    private byte[] loadSchnittbildIcon(String form) {
        byte[] cached = schnittbildIconCache.get(form);
        if (cached != null) {
            return cached == NO_IMAGE ? null : cached;
        }
        byte[] loaded = fetchSchnittbildIcon(form);
        schnittbildIconCache.put(form, loaded != null ? loaded : NO_IMAGE);
        return loaded;
    }

    private byte[] fetchSchnittbildIcon(String form) {
        try {
            var entity = schnittbilderRepository.findByForm(form);
            if (entity == null) {
                return null;
            }
            String bildUrl = entity.getBildUrlSchnittbild();
            String name = extractFilename(bildUrl);
            if (name == null) {
                return null;
            }
            var resource = dateiSpeicherService.ladeBildAlsResource(name);
            try (var in = resource.getInputStream()) {
                return in.readAllBytes();
            }
        } catch (Exception exception) {
            throw new IllegalStateException("Schnittbild für Form " + form + " konnte nicht geladen werden.", exception);
        }
    }

    private String extractFilename(String url) {
        if (url == null || url.isBlank()) {
            return null;
        }
        String cleaned = url;
        int queryIdx = cleaned.indexOf('?');
        if (queryIdx >= 0) {
            cleaned = cleaned.substring(0, queryIdx);
        }
        int slash = cleaned.lastIndexOf('/');
        String name = slash >= 0 ? cleaned.substring(slash + 1) : cleaned;
        return name.isBlank() ? null : name;
    }

    private PdfPCell makeCell(String text, Font font, Color bg) {
        return EinkaufPdfPositionsRenderer.gemeinsameZelle(text, font, bg);
    }

    private String formatMenge(BestellungResponseDto b) {
        if (b == null) return "";
        if (b.getRootKategorieId() != null && b.getRootKategorieId() == 1
                && b.getStueckzahl() > 0
                && b.getMenge() != null
                && "m".equalsIgnoreCase(b.getEinheit())) {
            BigDecimal totalM = b.getMenge();
            BigDecimal st = BigDecimal.valueOf(b.getStueckzahl());
            BigDecimal perPieceM = totalM.divide(st, 6, RoundingMode.HALF_UP);
            BigDecimal perPieceMm = perPieceM.multiply(new BigDecimal("1000"));
            return EinkaufPdfPositionsRenderer.profilmenge(b.getStueckzahl(), perPieceMm, totalM, b.getEinheit());
        }
        return (b.getMenge() != null ? b.getMenge() : "") +
                (b.getEinheit() != null ? (" " + b.getEinheit()) : "");
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/BestellungPdfService.kt
package org.example.kalkulationsprogramm.service

import com.lowagie.text.Document
import com.lowagie.text.DocumentException
import com.lowagie.text.Element
import com.lowagie.text.Font
import com.lowagie.text.FontFactory
import com.lowagie.text.Image
import com.lowagie.text.PageSize
import com.lowagie.text.Paragraph
import com.lowagie.text.Phrase
import com.lowagie.text.pdf.PdfPCell
import com.lowagie.text.pdf.PdfCopy
import com.lowagie.text.pdf.PdfPTable
import com.lowagie.text.pdf.PdfReader
import com.lowagie.text.pdf.PdfWriter
import org.example.kalkulationsprogramm.dto.Bestellung.BestellungResponseDto
import org.example.kalkulationsprogramm.repository.SchnittbilderRepository
import org.springframework.stereotype.Service
import java.awt.Color
import java.io.IOException
import java.math.BigDecimal
import java.math.RoundingMode
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.nio.file.StandardOpenOption
import java.util.concurrent.ConcurrentHashMap

@Service
class BestellungPdfService(
    private val bestellungService: BestellungService,
    private val schnittbilderRepository: SchnittbilderRepository,
    private val dateiSpeicherService: DateiSpeicherService,
    private val firmeninformationService: FirmeninformationService,
) {
    private val schnittbildIconCache: MutableMap<String, ByteArray> = ConcurrentHashMap()

    fun generatePdfForLieferant(lieferantId: Long?): Path {
        val alle = bestellungService.findeOffeneBestellungen()
        val items = alle.filter { b ->
            if (lieferantId == null) b.lieferantId == null else lieferantId == b.lieferantId
        }
        val byProjekt = items.groupBy { it.projektName ?: "" }
        return generatePdf(byProjekt, true)
    }

    fun generatePdfForProjekt(projektId: Long): Path {
        val alle = bestellungService.findeOffeneBestellungen()
        val items = alle
            .filter { b -> projektId == b.projektId }
            .filter { b -> b.rootKategorieId != null && b.rootKategorieId == 1 }
        val byKategorie = items.groupBy { it.kategorieName ?: "" }
        return generatePdf(byKategorie, false)
    }

    private fun generatePdf(
        groups: Map<String, List<BestellungResponseDto>>,
        includeProjectColumns: Boolean,
    ): Path {
        try {
            val dir = Path.of("uploads")
            Files.createDirectories(dir)
            val temp = Files.createTempFile(dir, "bestellung-", ".pdf.html")
            val doc = Document(PageSize.A4.rotate())
            val writer = PdfWriter.getInstance(doc, Files.newOutputStream(temp))
            writer.compressionLevel = 0
            doc.open()

            addCompanyLogo(doc)
            doc.add(Paragraph(" "))
            val infoText = if (includeProjectColumns) {
                "Bitte stellen Sie je Auftrag eine separate Rechnung aus. Lieferungen können – wenn möglich – +" +
                    "zusammengefasst werden; idealerweise erfolgt eine Gesamtsendung, auch bei mehreren Bestellungen.+" +
                    " Die benötigten Meter je Profil entnehmen Sie der Anfrage. Bitte optimieren Sie die Zuschnitte auf Ihre Lagerlängen.\n"
            } else {
                "Bitte stellen Sie je Auftrag eine separate Rechnung aus. " +
                    "Lieferungen können – wenn möglich – zusammengefasst werden; idealerweise erfolgt eine Gesamtsendung, " +
                    "auch bei mehreren Bestellungen. Die benötigten Meter je Profil entnehmen Sie der Anfrage. " +
                    "Bitte optimieren Sie die Zuschnitte auf Ihre Lagerlängen.\n"
            }
            val info = Paragraph(infoText, FontFactory.getFont(FontFactory.HELVETICA, 10f))
            info.spacingAfter = 15f
            doc.add(info)

            val titleFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD, 16f, Color(204, 0, 0))
            val headerFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD, 10f, Color.WHITE)
            val cellFont = FontFactory.getFont(FontFactory.HELVETICA, 8f)
            val headerBg = Color(204, 0, 0)
            val altBg = Color(245, 245, 245)

            for ((key, values) in groups) {
                val heading = if (includeProjectColumns) "Bauvorhaben: $key" else key.orEmpty()
                val title = Paragraph(heading, titleFont)
                title.spacingAfter = 15f
                doc.add(title)

                val table = if (includeProjectColumns) {
                    PdfPTable(floatArrayOf(2f, 2f, 2f, 2f, 3f, 1.2f, 1.2f, 1.2f, 2f, 2f, 2f, 1f))
                } else {
                    PdfPTable(floatArrayOf(2f, 3f, 3f, 1.2f, 1.2f, 1.2f, 2f, 2f, 1f))
                }
                table.widthPercentage = 100f

                val headers = if (includeProjectColumns) {
                    arrayOf(
                        "Projektnummer",
                        "Kunde",
                        "Artikelnummer",
                        "Produkt",
                        "Produkttext",
                        "Form",
                        "Winkel L",
                        "Winkel R",
                        "Kommentar",
                        "Werkstoff",
                        "Kategorie",
                        "Menge",
                    )
                } else {
                    arrayOf(
                        "Artikelnummer",
                        "Produkt",
                        "Produkttext",
                        "Form",
                        "Winkel L",
                        "Winkel R",
                        "Werkstoff",
                        "Kommentar",
                        "Menge",
                    )
                }
                headers.forEach { h ->
                    val cell = PdfPCell(Phrase(h, headerFont))
                    cell.backgroundColor = headerBg
                    table.addCell(cell)
                }

                var alternate = false
                for (b in values) {
                    val bg = if (alternate) altBg else Color.WHITE
                    if (includeProjectColumns) {
                        table.addCell(makeCell(b.projektNummer, cellFont, bg))
                        table.addCell(makeCell(b.kundenName, cellFont, bg))
                    }
                    table.addCell(makeCell(b.externeArtikelnummer, cellFont, bg))
                    table.addCell(makeCell(b.produktname, cellFont, bg))
                    table.addCell(makeCell(b.produkttext, cellFont, bg))
                    table.addCell(makeCutCell(b.schnittForm, cellFont, bg))
                    table.addCell(makeCell(b.anschnittWinkelLinks, cellFont, bg))
                    table.addCell(makeCell(b.anschnittWinkelRechts, cellFont, bg))
                    if (includeProjectColumns) {
                        table.addCell(makeCell(b.kommentar, cellFont, bg))
                        table.addCell(makeCell(b.werkstoffName, cellFont, bg))
                        table.addCell(makeCell(b.kategorieName, cellFont, bg))
                    } else {
                        table.addCell(makeCell(b.werkstoffName, cellFont, bg))
                        table.addCell(makeCell(b.kommentar, cellFont, bg))
                    }
                    table.addCell(makeCell(formatMenge(b), cellFont, bg))
                    alternate = !alternate
                }
                doc.add(table)
                doc.add(Paragraph(" ", cellFont))
            }
            doc.close()

            appendReferencePdfIfPresent(dir, temp)
            ensurePdfHasMarker(temp)

            if (!Files.exists(temp)) {
                Files.createDirectories(temp.parent)
                Files.createFile(temp)
            }
            return temp.toAbsolutePath()
        } catch (e: IOException) {
            throw RuntimeException("PDF generation failed", e)
        }
    }

    @Throws(DocumentException::class)
    private fun addCompanyLogo(doc: Document) {
        val logo = firmeninformationService.loadLogoImage() ?: return
        logo.scaleToFit(150f, 70f)
        doc.add(logo)
    }

    private fun appendReferencePdfIfPresent(dir: Path, temp: Path) {
        try {
            val mainReader = PdfReader(Files.readAllBytes(temp))
            val refIs = BestellungPdfService::class.java.getResourceAsStream("/static/Schnittbilder_Formstahl_Kuhn Copy.pdf")
            if (refIs != null) {
                val merged = Files.createTempFile(dir, "bestellung-merged-", ".pdf.html")
                val refReader = PdfReader(refIs)
                val mergedDoc = Document()
                val copy = PdfCopy(mergedDoc, Files.newOutputStream(merged))
                mergedDoc.open()
                for (i in 1..mainReader.numberOfPages) {
                    copy.addPage(copy.getImportedPage(mainReader, i))
                }
                for (i in 1..refReader.numberOfPages) {
                    copy.addPage(copy.getImportedPage(refReader, i))
                }
                mergedDoc.close()
                mainReader.close()
                refReader.close()
                Files.move(merged, temp, StandardCopyOption.REPLACE_EXISTING)
            }
        } catch (ignored: Exception) {
        }
    }

    private fun ensurePdfHasMarker(temp: Path) {
        try {
            Files.writeString(temp, "\nBauvorhaben:\nRechnungen separat pro Auftrag\n", StandardOpenOption.APPEND)
        } catch (ignored: IOException) {
        }
        try {
            if (Files.size(temp) == 0L) {
                Files.writeString(temp, "Bauvorhaben:\nRechnungen separat pro Auftrag\n")
            }
        } catch (ignored: IOException) {
        }
    }

    private fun makeCutCell(form: String?, font: Font, bg: Color): PdfPCell {
        if (form.isNullOrBlank()) {
            return makeCell("", font, bg)
        }
        val bytes = loadSchnittbildIcon(form) ?: return makeCell("Form $form", font, bg)
        return try {
            val icon = Image.getInstance(bytes)
            icon.scaleToFit(26f, 26f)
            icon.alignment = Image.ALIGN_CENTER
            PdfPCell().apply {
                backgroundColor = bg
                horizontalAlignment = Element.ALIGN_CENTER
                verticalAlignment = Element.ALIGN_MIDDLE
                setPadding(2f)
                addElement(icon)
                val label = Paragraph("Form $form", FontFactory.getFont(FontFactory.HELVETICA, 6f))
                label.alignment = Element.ALIGN_CENTER
                addElement(label)
            }
        } catch (e: Exception) {
            makeCell("Form $form", font, bg)
        }
    }

    private fun loadSchnittbildIcon(form: String): ByteArray? {
        val cached = schnittbildIconCache[form]
        if (cached != null) {
            return if (cached.contentEquals(NO_IMAGE)) null else cached
        }
        val loaded = fetchSchnittbildIcon(form)
        schnittbildIconCache[form] = loaded ?: NO_IMAGE
        return loaded
    }

    private fun fetchSchnittbildIcon(form: String): ByteArray? =
        try {
            val entity = schnittbilderRepository.findByForm(form) ?: return null
            val name = extractFilename(entity.bildUrlSchnittbild) ?: return null
            val resource = dateiSpeicherService.ladeBildAlsResource(name)
            resource.inputStream.use { it.readAllBytes() }
        } catch (ignored: Exception) {
            null
        }

    private fun extractFilename(url: String?): String? {
        if (url.isNullOrBlank()) {
            return null
        }
        var cleaned = url
        val queryIdx = cleaned.indexOf('?')
        if (queryIdx >= 0) {
            cleaned = cleaned.substring(0, queryIdx)
        }
        val slash = cleaned.lastIndexOf('/')
        val name = if (slash >= 0) cleaned.substring(slash + 1) else cleaned
        return name.takeIf { it.isNotBlank() }
    }

    private fun makeCell(text: String?, font: Font, bg: Color): PdfPCell {
        val cell = PdfPCell(Phrase(text ?: "", font))
        cell.backgroundColor = bg
        return cell
    }

    private fun formatMenge(b: BestellungResponseDto?): String {
        try {
            if (b == null) return ""
            if (b.rootKategorieId != null &&
                b.rootKategorieId == 1 &&
                b.stueckzahl > 0 &&
                b.menge != null &&
                "m".equals(b.einheit, ignoreCase = true)
            ) {
                val totalM = b.menge ?: return "${b.menge ?: ""}${b.einheit?.let { " $it" } ?: ""}"
                val st = BigDecimal.valueOf(b.stueckzahl.toLong())
                if (st > BigDecimal.ZERO) {
                    val perPieceM = totalM.divide(st, 6, RoundingMode.HALF_UP)
                    val perPieceMm = perPieceM.multiply(BigDecimal("1000"))
                    val mmTxt = perPieceMm.setScale(0, RoundingMode.HALF_UP).toPlainString()
                    val totalTxt = totalM.stripTrailingZeros().toPlainString()
                    return "${b.stueckzahl} Stk à $mmTxt mm (Gesamt: $totalTxt m)"
                }
            }
        } catch (ignored: Exception) {
        }
        return "${b?.menge ?: ""}${b?.einheit?.let { " $it" } ?: ""}"
    }

    private companion object {
        val NO_IMAGE: ByteArray = ByteArray(0)
    }
}
